Overhaul of the documentation for new comers

Thomsch opened the pull request:
The objective of these changes is to provide a more precise overview of what Oríon is and what its features are. 

This is achieved through:
- a refresh of the README, with additional links to resources
- the addition of an overview for the user guide
- an update of the installation instructions
- the addition of a getting started guide, providing users a comprehensive tour of Oríon with relevant links.

Please share your suggestions to improve this documentation: What would you like to see that's not here? What could be improved?  Name it!

Note that the database setup will also be rewritten later in another PR to complement these changes

Thomsch commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Révision de la documentation dans une perspective d'un nouvel utilisateur.](https://trello.com/c/d7xYD5it/397-r%C3%A9vision-de-la-documentation-dans-une-perspective-dun-nouvel-utilisateur)

bouthilx commented on README.rst:
```diff
@@ -48,8 +48,12 @@ now what you have to do looks like this:
 
-Check out our `pytorch example`_ for a quick demonstration!
+Check out our `getting started guide`_ or `this presentation
+<https://docs.google.com/presentation/d/18g7Q4xRuhMtcVbwmFwDfH7v9gKS252-laOi9HrEQ7a4/edit?usp=sharing>`_
```
I'm not sure the slides are easy to follow without explanations. I tend to make minimalist slides without much text. Do we have access to the video recording outside of Mila instead?

Thomsch replied:
We do it's here: https://bluejeans.com/playback/s/4WUezzFCmb9StHzYgB0RjVbTUCKnRcptBvzBMP7t2UpLyKuAq7Emieo911BqEMnI

bouthilx commented on docs/src/install/core.rst:
```diff
@@ -2,31 +2,34 @@
-
-The easiest way to install Oríon is using the Python package manager. The core of Oríon is
-registered on PyPI under `orion`.
+The easiest way to install the latest version of Oríon is through the Python package manager. Oríon
```
This is a bit confusing because `latest` in docs now points to develop branch while stable points to `master`, which is what you get using PyPI.

Thomsch replied:
Do you mean that on PyPI, `latest` means the last version to have been released? If it's the case we have to either roll with it as the `latest` on PyPI isn't related to our `latest` readthedoc channel or we have to remove the `stable` channel; we can't have 2 channels pointing to the same thing, it's even more confusing. 
Which alternative do you prefer?

bouthilx replied:
Develop is not available on PyPI, so yes latest would be the equivalent of master. We need to have access to develop version on readthedocs. Maybe latest and develop, or stable and develop?

Thomsch replied:
That's was I was thinking about too as a solution. I prefer having `stable` and `develop` because these channel names are more descriptive and `stable` channel is a staple in readthedocs ecosystem.

bouthilx commented on docs/src/install/core.rst:
```diff
@@ -2,31 +2,34 @@
 
-This will install all the core components. Note that the only algorithm provided with it
-is random search. To install more algorithms, you can look at section :doc:`/plugins/install`.
+Note that Oríon comes with the following algorithms: Random Search, Hyperband, and ASHA. More
```
You can now add TPE. :slightly_smiling_face:  :tada:

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+streamline your machine learning workflow whether you are a researcher or engineer.
+
+Oríon is a black box function optimization library with a key focus on usability and integrability
+for its users. For example, as an machine learning engineer, you can integrate Oríon to your
```
```suggestion
for its users. For example, as a machine learning engineer, you can integrate Oríon to your
```

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+conditions.
+
+Oríon is built to be non-intrusive, highly integrable, operate in parallel environments, and produce
+reproducible experiments. It takes only one line of code to start using it in your existing
```
Maybe at 'if your script support hyperparameter definition through commandline arguments of configuration files, then it takes only one line of code ...'.

Thomsch replied:
This is a good suggestion but I feel it sounds too specific for an introduction.

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+not the case please refer to our :doc:`installation instructions </install/core>` and :doc:`database
+setup </install/database>`.
+
+We also made a presentation_ if you prefer easy-to-read slides!
```
> easy-to-read slides
Are they? :sweat_smile:

Thomsch replied:
They are! They have a good balance of information which is great :). Nonetheless, I replaced it with the bluejeans presentation. It's always nicer to have someone explain the slides for you!

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+encourage developers to integrate Oríon into that kind of systems as a component and we will do
+our best to help you if you're interested.
+
+Before continuing the overview, we assume that you have a basic understanding of machine learning
```
This breaks the flow. I think the rest of this page is readable without making the installation and db setup.

Thomsch replied:
Where would you put it? If feel it's good to have it on this page, otherwise we never explicitly say to install Oríon and the database except in the README. Scikit-learn's getting started is using this model also.

bouthilx replied:
I don't think it is necessary to have installed Orion and setup the dataset to understand the overview, but I get your point about where should we point out doc for installation. Maybe changing a bit the wording would make it sound less mandatory. Something like the following.

> Before continuing the overview, we assume that you have a basic understanding of machine learning concepts. You may also want to install Oríon on your machine and configure it for a database before continuing. Please refer to our :doc:`installation instructions </install/core>` and :doc:`database
setup </install/database>`.

Thomsch replied:
I like it!

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+trial will be stored in the database that you configured during the installation process.
+
+You can fine-tune the distribution and algorithm with many options either with more arguments or by
+using a configuration file. Learn more at `/user/api`.
```
Link is broken here.

Thomsch replied:
Good catch!

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+   $ orion hunt -n <experiment name> script.py --lr~'loguniform(1e-5, 1.0)'
+
+This is going to start the optimization process using the default optimization algorithm and sample
+the values for the **lr** hyper-parameter in a log uniform distribution between 0.00001 et 1.0. Each
```
```suggestion
the values for the ``lr`` hyper-parameter in a log uniform distribution between 0.00001 et 1.0. Each
```
Do you think that would be more readable?

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+the database, each worker will separately generate a new trial based on the state of the experiment
+stored in the database.
+
+Make sure to visit :doc:`/user/parallel` to learn more about it and checkout the tutorial to run
```
```suggestion
Make sure to visit :doc:`/user/parallel` to learn more about it and check out the tutorial to run
```

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+==========
+
+Similarly to search spaces, Oríon supports multiple algorithms out of the box: :ref:`random-search`,
+:ref:`ASHA`, and :ref:`hyperband-algorithm`. Each one is fully configurable through the
```
And now TPE! :)

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+
+Make sure to checkout `this presentation
+<https://docs.google.com/presentation/d/18g7Q4xRuhMtcVbwmFwDfH7v9gKS252-laOi9HrEQ7a4/present?slide=id.g6ba6d709b9_4_19>`_
+for an quick overview of each algorithm and to visit :doc:`/user/algorithms` to learn about the
```
```suggestion
for a quick overview of each algorithm and to visit :doc:`/user/algorithms` to learn about the
```

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+It's worth to take a look at the :doc:`configuration system </user/config>` to learn more about how
+to make the most out of Oríon and define precise behaviors for your algorithms and experiments.
+
+Explore the :doc:`User Manual </user/overview>`, oríon is simple from the outside but is feature
```
```suggestion
Explore the :doc:`User Manual </user/overview>`, Oríon is simple from the outside but is feature
```

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+to make the most out of Oríon and define precise behaviors for your algorithms and experiments.
+
+Explore the :doc:`User Manual </user/overview>`, oríon is simple from the outside but is feature
+rich! We also have a few tutorials available. If you're a researcher or develop you might be
```
Maybe turn the word tutorial into a link pointing to simple example?

Thomsch replied:
I wanted to add a list of tutorials at first but we don't have such thing yet so instead, I'm adding a few examples.

bouthilx commented on docs/src/install/gettingstarted.rst:
```diff
@@ -0,0 +1,149 @@
+to make the most out of Oríon and define precise behaviors for your algorithms and experiments.
+
+Explore the :doc:`User Manual </user/overview>`, oríon is simple from the outside but is feature
+rich! We also have a few tutorials available. If you're a researcher or develop you might be
```
```suggestion
rich! We also have a few tutorials available. If you're a researcher or developper you might be
```

bouthilx commented on docs/src/user/overview.rst:
```diff
@@ -0,0 +1,26 @@
+
+The user guide is organized in the following sections:
+
+* :doc:`/user/script`. Describes how you can integrate Oríon to your existing scripts and
```
Maybe drop all the describes and start with 'How [...]'.
I wounder if using `:attr: desc.` like in [config documentation](https://raw.githubusercontent.com/Epistimio/orion/develop/docs/src/user/config.rst) would give a better layout.

Thomsch replied:
I tried to remove the *Describes* but it's a bit awkward without them.
I also tried the attributes but it takes way too much space for a simple index. It would have been fine if the bold meta-title could be a link but unfortunately, restructuredText doesn't allow it :(.

bouthilx commented on docs/src/user/parallel.rst:
```diff
@@ -0,0 +1,32 @@
+  :figclass: align-center
+
+If we spawn more worker for the experiment, their workflow is unchanged because the workers are
+synchronized during the creation of a new trial based on what other trials were already completed.
```
It may be worth showing explicitly that the user only needs to run multiple times the command above to spawn the new workers.

Thomsch replied:
Yes, of course!

bouthilx commented:
I left a couple of minor comments but overall this is great! This new getting started section should help a lot for new comers!

bouthilx commented:
Maybe the simple example could be renamed to pytorch-mnist and pushed below the scikit-learn example which is more generally accessible.

Thomsch commented:
I reworked the introduction. Specifically, I removed the unreadable chunk of text listing all the features of Oríon and moved it to the relevant sections on the page.

bouthilx commented on docs/src/tutorials/pytorch-mnist.rst:
```diff
@@ -1,5 +1,5 @@
 **************
```
```suggestion
*************
```

bouthilx commented on docs/src/tutorials/pytorch-mnist.rst:
```diff
@@ -1,5 +1,5 @@
 **************
-Simple example
+PyTorch MNIST
 **************
```
```suggestion
*************
```

bouthilx commented on docs/src/index.rst:
```diff
@@ -30,9 +30,9 @@
 
    tutorials/scikit-learn
    tutorials/pytorch-mnist
+   tutorials/cluster
```
I agree it's more important but there seems to be an implicit categorization. Like specific applications vs ways to use it. Do you think we should make it explicit and making subsections in the tutorial menu?

Thomsch replied:
I don't think it's worth it to do it at this time because there is not that many tutorials in total and because the two tutorials after *cluster* are very specific case.
